File: backend/app/models/schemas/kb_chat.py
"""Knowledge-base chat related schemas."""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class KbChatRequest(BaseModel):
    """Knowledge-base chat request."""

    query: str = Field(..., description="User question")
    team_id: Optional[int] = Field(None, description="Limit retrieval to one team")
    knowledge_base_id: Optional[int] = Field(None, description="Limit retrieval to one KB")
    category_id: Optional[int] = Field(None, description="Limit retrieval to one category")
    assistant_id: Optional[int] = Field(None, description="Assistant profile id")
    session_id: Optional[str] = Field(None, description="Optional session id")


class KbChatResponse(BaseModel):
    """Knowledge-base chat response."""

    answer: str = Field(..., description="Answer content")
    answer_text: str = Field(..., description="Answer content for end-user rendering")
    answer_status: str = Field("answered", description="Answer confidence/result state")
    confidence_level: str | None = Field(None, description="Optional confidence band")
    backend_citations: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Retrieved citations retained for admin QA",
    )
    retrieved_docs: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Retrieved document chunks",
    )
    assistant_id: Optional[int] = None
    assistant_name: Optional[str] = None
    session_id: Optional[str] = None
    log_id: Optional[int] = None


class KbChatSessionMessage(BaseModel):
    """One persisted message in a KB chat session."""

    role: str = Field(..., description="Message role")
    content: str = Field(..., description="Message content")
    retrieved_docs: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="Retrieved document chunks for one assistant message",
    )
    answer_status: Optional[str] = Field(None, description="Persisted answer state")
    log_id: Optional[int] = Field(None, description="Persisted KB chat log id")
    created_at: datetime = Field(..., description="Creation timestamp")


class KbChatSessionSummary(BaseModel):
    """Summary item for a user's persisted KB chat session."""

    session_id: str = Field(..., description="Stable session identifier")
    title: str = Field(..., description="Display title derived from the conversation")
    preview: Optional[str] = Field(None, description="Short preview of the latest message")
    team_id: Optional[int] = Field(None, description="Selected team id")
    knowledge_base_id: Optional[int] = Field(None, description="Selected knowledge base id")
    knowledge_base_name: Optional[str] = Field(None, description="Selected knowledge base name")
    assistant_id: Optional[int] = Field(None, description="Selected assistant id")
    assistant_name: Optional[str] = Field(None, description="Selected assistant name")
    category_id: Optional[int] = Field(None, description="Selected category id")
    category_name: Optional[str] = Field(None, description="Selected category name")
    message_count: int = Field(..., description="Persisted message count")
    created_at: datetime = Field(..., description="Session creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")


class KbChatSessionDetail(KbChatSessionSummary):
    """Full detail for a persisted KB chat session."""

    messages: List[KbChatSessionMessage] = Field(
        default_factory=list,
        description="Persisted chat messages in chronological order",
    )


class KbChatPreviewRequest(KbChatRequest):
    """Admin-only KB preview request that can read unpublished content."""

    include_unpublished: bool = Field(
        default=True,
        description="Whether to include draft/indexed content in retrieval",
    )


class KbChatFeedbackRequest(BaseModel):
    """User feedback for one KB chat log entry."""

    feedback_value: str = Field(..., min_length=1, max_length=20)
    feedback_note: str | None = Field(default=None, max_length=1000)


class KbChatReviewRequest(BaseModel):
    """Admin review label and note for one KB chat log entry."""

    review_label: str | None = Field(default=None, max_length=40)
    review_note: str | None = Field(default=None, max_length=1000)


class KbChatLogItem(BaseModel):
    """Admin-facing KB chat log item."""

    id: int
    user_id: int
    session_id: str | None = None
    team_id: int | None = None
    team_name: str | None = None
    knowledge_base_id: int | None = None
    knowledge_base_name: str | None = None
    assistant_id: int | None = None
    assistant_name: str | None = None
    category_id: int | None = None
    category_name: str | None = None
    query: str
    answer_text: str
    answer_status: str
    retrieval_status: str | None = None
    retrieved_count: int = 0
    latency_ms: int | None = None
    feedback_value: str | None = None
    feedback_note: str | None = None
    suggested_review_label: str | None = None
    review_label: str | None = None
    review_note: str | None = None
    reviewed_at: datetime | None = None
    reviewed_by_user_id: int | None = None
    created_at: datetime


class KbChatDiagnosticDoc(BaseModel):
    """One retrieved chunk retained for QA diagnostics."""

    rank: int = 0
    content: str = ""
    metadata: Dict[str, Any] = Field(default_factory=dict)


class KbChatDiagnosticMessage(BaseModel):
    """One surrounding message for diagnostic context."""

    role: str
    content: str
    created_at: datetime
    is_current_turn: bool = False


class KbChatRetrievalQueryStat(BaseModel):
    """One rewritten retrieval query and its recall count."""

    query: str
    chunk_count: int = 0


class KbChatRetrievalFunnelStage(BaseModel):
    """One stage in the retrieval funnel."""

    key: str
    label: str
    chunk_count: int = 0
    note: str | None = None


class KbChatRetrievalFunnel(BaseModel):
    """Structured retrieval funnel retained for diagnostics."""

    mode: str | None = None
    query_count: int = 0
    rewritten_queries: List[KbChatRetrievalQueryStat] = Field(default_factory=list)
    stages: List[KbChatRetrievalFunnelStage] = Field(default_factory=list)


class KbChatLogDetail(KbChatLogItem):
    """Admin-facing detailed diagnostic view for one KB chat log."""

    team_id: int | None = None
    team_name: str | None = None
    assistant_id: int | None = None
    assistant_name: str | None = None
    category_name: str | None = None
    retrieval_status_reason: str | None = None
    retrieval_queries: List[str] = Field(default_factory=list)
    retrieval_funnel: KbChatRetrievalFunnel | None = None
    answer_context: str | None = None
    retrieved_docs: List[KbChatDiagnosticDoc] = Field(default_factory=list)
    conversation_context: List[KbChatDiagnosticMessage] = Field(default_factory=list)


class KbChatLogListResponse(BaseModel):
    """Admin-facing KB chat log list response."""

    items: List[KbChatLogItem] = Field(default_factory=list)
    total: int = 0

File: frontend/lib/api/endpoints/kbChat.ts
import { apiClient } from "../client";

export interface KbChatRequest {
  query: string;
  team_id?: number | null;
  knowledge_base_id?: number | null;
  category_id?: number | null;
  session_id?: string | null;
}

export interface AskTeamOption {
  id: number;
  name: string;
  code?: string | null;
  description?: string | null;
  created_at: string;
  updated_at: string;
}

export interface RetrievedDoc {
  content: string;
  metadata?: {
    document_id?: number;
    document_title?: string;
    chunk_index?: number;
    score?: number;
    rerank_score?: number;
    category_id?: number | null;
    category_name?: string | null;
    source_path?: string | null;
  };
}

export interface KbChatResponse {
  answer: string;
  answer_text: string;
  answer_status: string;
  confidence_level?: string | null;
  backend_citations: RetrievedDoc[];
  retrieved_docs: RetrievedDoc[];
  session_id?: string | null;
  log_id?: number | null;
}

export interface KbChatSessionMessage {
  role: string;
  content: string;
  retrieved_docs?: RetrievedDoc[];
  answer_status?: string | null;
  log_id?: number | null;
  created_at: string;
}

export interface KbChatSessionSummary {
  session_id: string;
  title: string;
  preview?: string | null;
  team_id?: number | null;
  knowledge_base_id?: number | null;
  knowledge_base_name?: string | null;
  category_id?: number | null;
  category_name?: string | null;
  message_count: number;
  created_at: string;
  updated_at: string;
}

export interface KbChatSessionDetail extends KbChatSessionSummary {
  messages: KbChatSessionMessage[];
}

export interface KbChatLogItem {
  id: number;
  user_id: number;
  session_id?: string | null;
  team_id?: number | null;
  team_name?: string | null;
  knowledge_base_id?: number | null;
  knowledge_base_name?: string | null;
  category_id?: number | null;
  category_name?: string | null;
  query: string;
  answer_text: string;
  answer_status: string;
  retrieval_status?: string | null;
  retrieved_count: number;
  latency_ms?: number | null;
  feedback_value?: string | null;
  feedback_note?: string | null;
  suggested_review_label?: string | null;
  review_label?: string | null;
  review_note?: string | null;
  reviewed_at?: string | null;
  reviewed_by_user_id?: number | null;
  created_at: string;
}

export interface KbChatDiagnosticDoc {
  rank: number;
  content: string;
  metadata: {
    document_id?: number;
    document_title?: string;
    chunk_index?: number;
    score?: number;
    rerank_score?: number;
    category_id?: number | null;
    category_name?: string | null;
    source_path?: string | null;
  };
}

export interface KbChatDiagnosticMessage {
  role: string;
  content: string;
  created_at: string;
  is_current_turn: boolean;
}

export interface KbChatRetrievalQueryStat {
  query: string;
  chunk_count: number;
}

export interface KbChatRetrievalFunnelStage {
  key: string;
  label: string;
  chunk_count: number;
  note?: string | null;
}

export interface KbChatRetrievalFunnel {
  mode?: string | null;
  query_count: number;
  rewritten_queries: KbChatRetrievalQueryStat[];
  stages: KbChatRetrievalFunnelStage[];
}

export interface KbChatLogDetail extends KbChatLogItem {
  team_id?: number | null;
  team_name?: string | null;
  category_name?: string | null;
  retrieval_status_reason?: string | null;
  retrieval_queries: string[];
  retrieval_funnel?: KbChatRetrievalFunnel | null;
  answer_context?: string | null;
  retrieved_docs: KbChatDiagnosticDoc[];
  conversation_context: KbChatDiagnosticMessage[];
}

export interface KbChatLogListResponse {
  items: KbChatLogItem[];
  total: number;
}

export interface KbChatLogListFilters {
  limit?: number;
  team_id?: number | null;
  knowledge_base_id?: number | null;
  category_id?: number | null;
  answer_status?: string | null;
  retrieval_status?: string | null;
  feedback_value?: string | null;
  query_keyword?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  zero_hits_only?: boolean;
  high_latency_only?: boolean;
  high_latency_threshold_ms?: number;
}

export const kbChatApi = {
  invoke: (request: KbChatRequest) =>
    apiClient.post<KbChatResponse>("/api/v1/ask/invoke", request),

  stream: (request: KbChatRequest) =>
    apiClient.postStream("/api/v1/ask/stream", request),

  listSessions: (limit: number = 30) =>
    apiClient.get<KbChatSessionSummary[]>(`/api/v1/ask/sessions?limit=${limit}`),

  listTeams: () => apiClient.get<AskTeamOption[]>("/api/v1/ask/teams"),

  getSession: (sessionId: string) =>
    apiClient.get<KbChatSessionDetail>(
      `/api/v1/ask/sessions/${encodeURIComponent(sessionId)}`,
    ),

  deleteSession: (sessionId: string) =>
    apiClient.delete<void>(`/api/v1/ask/sessions/${encodeURIComponent(sessionId)}`),

  submitFeedback: (
    logId: number,
    payload: { feedback_value: string; feedback_note?: string | null },
  ) => apiClient.post(`/api/v1/ask/feedback/${logId}`, payload),

  reviewLog: (
    logId: number,
    payload: { review_label?: string | null; review_note?: string | null },
  ) => apiClient.post(`/api/v1/admin/qa/logs/${logId}/review`, payload),

  preview: (request: KbChatRequest & { include_unpublished?: boolean }) =>
    apiClient.post<KbChatResponse>("/api/v1/admin/qa/preview", request),

  listLogs: (filters: KbChatLogListFilters = {}) => {
    const params = new URLSearchParams();
    params.set("limit", String(filters.limit ?? 50));
    if (filters.team_id != null) params.set("team_id", String(filters.team_id));
    if (filters.knowledge_base_id != null) {
      params.set("knowledge_base_id", String(filters.knowledge_base_id));
    }
    if (filters.category_id != null) params.set("category_id", String(filters.category_id));
    if (filters.answer_status) params.set("answer_status", filters.answer_status);
    if (filters.retrieval_status) params.set("retrieval_status", filters.retrieval_status);
    if (filters.feedback_value) params.set("feedback_value", filters.feedback_value);
    if (filters.query_keyword) params.set("query_keyword", filters.query_keyword);
    if (filters.start_date) params.set("start_date", filters.start_date);
    if (filters.end_date) params.set("end_date", filters.end_date);
    if (filters.zero_hits_only) params.set("zero_hits_only", "true");
    if (filters.high_latency_only) params.set("high_latency_only", "true");
    if (filters.high_latency_threshold_ms != null) {
      params.set("high_latency_threshold_ms", String(filters.high_latency_threshold_ms));
    }
    return apiClient.get<KbChatLogListResponse>(`/api/v1/admin/qa/logs?${params.toString()}`);
  },

  getLogDetail: (logId: number) =>
    apiClient.get<KbChatLogDetail>(`/api/v1/admin/qa/logs/${logId}`),
};
